### DISCUSSION TITLE
Custom Object: caching, controls, clipping, bbox

### DISCUSSION THREAD

**Discussion by umtdemr** [fabricjs/fabric.js#8567] [Q&A] [answered]
Hi, I created a new fabric js class from fabric.Polyline class. The thing is, I want to hide some parts of the line using `ctx.clearRect` in _render method. But setting objectCaching to false breaks clearRect. It doesn't clear the part of the line, it just draws a white rectangle. Since I need to control points of the line with corners, the objectCaching should be false in this line. Is there any way to fix this without setting objectCaching to true?

I tried to achieve the same thing with `ctx.globalCompositeOperation = 'destination-out'`  but the results were the same.

js fiddle: [https://jsfiddle.net/mediumgoal/sn5pbrxt/36/](https://jsfiddle.net/mediumgoal/sn5pbrxt/36/)

**Comment by ShaMan123** [comment, marked as answer]
Why not use objectCaching?
You are in need of context isolation in order to clip the shape so you must use a cache canvas.
Simply invalidate it properly...
I suggest using clipPath instead of clipping manually...

**Reply by umtdemr** [reply]
@ShaMan123 but I'm changing the position of the line points with controls. It's like this example: [http://fabricjs.com/custom-controls-polygon](http://fabricjs.com/custom-controls-polygon). In order to do this, settiing objectCaching to false is required. Otherwise, it behaves strangely.

**Reply by ShaMan123** [reply]
simply mark it as dirty each step

**Reply by umtdemr** [reply]
I set `objectCaching` to false and as you said, I set `dirty`to true for the line in every step but it acts weirdly. The same weird thing happens on [fabric js's custom controls polygon example](http://fabricjs.com/custom-controls-polygon) when I set the same properties.

https://user-images.githubusercontent.com/61540809/210630083-52212682-ba96-4803-9352-f6428df74934.mov

**Reply by ShaMan123** [reply]
set object caching to true you mean?
what you see is caching artifacts not being cleaned up properly.

**Reply by umtdemr** [reply]
Sorry, yes you're right. I mean setting object caching to true. 

> what you see is caching artifacts not being cleaned up properly.

I see... is there any way to solve it?

**Reply by ShaMan123** [reply]
it means you are not invalidating caches properly.
You should mark as dirty from within the actionHandler

https://user-images.githubusercontent.com/34343793/210631875-f34637a7-2062-4dea-8fc2-560a2182b34e.mp4

**Reply by ShaMan123** [reply]
I am guessing you a rendering shapes as part of a container so make sure you control caching at all levels

**Reply by umtdemr** [reply]
> I am guessing you a rendering shapes as part of a container so make sure you control caching at all levels

Actually in my example there is only one element. I'm rendering the text inside of it with `ctx.fillText`.

> You should mark as dirty from within the actionHandler

I'm setting dirty to true in actionHandler but the result is similar to old one.

**Reply by ShaMan123** [reply]
Sorry, w/o a repro I can't do any better

**Reply by ShaMan123** [reply]
But if you are an early adopter then v6 can make things easier. Group in v6 is designed for all this beauty.

**Reply by umtdemr** [reply]
> Sorry, w/o a repro I can't do any better

Here is the exact custom class I'm working on: [https://jsfiddle.net/mediumgoal/hpkcyj63/31//](https://jsfiddle.net/mediumgoal/hpkcyj63/31/)

> But if you are an early adopter then v6 can make things easier. Group in v6 is designed for all this beauty.

I can give it a shot. How can I be an early adopter?

**Reply by ShaMan123** [reply]
> I can give it a shot. How can I be an early adopter?

You install fabric directly from github and run `npm run build` from the fabric folder.

One major thing stands out when looking at your example. 
Polyline wasn't designed for curves so you need to adapt the bounding box calculation.
Meaning that you should override the `_calcDimensions` method.
Try using `getBoundsOfCurve` to add that to the bounding box provided by  `_calcDimensions`.
I think that will solve the issue.

**Reply by umtdemr** [reply]
Okay, I'll try to override `_calcDimensions` using `getBoundsOfCurve`.

> Polyline wasn't designed for curves so you need to adapt the bounding box calculation.

I know that actually. Does fabric js provide any class for quadratic curves except `fabric.Path`?

**Reply by ShaMan123** [reply]
Not close to perfect... but gives you a starting point. I guessed most of it so don't take it as correct.
Might be that cpX/Y need pathOffset in _calcDimensions

https://jsfiddle.net/ks7rfzac/

```
  _calcDimensions() {
    this.pathOffset = this.pathOffset || new fabric.Point(0, 0)
    const x = fabric.Polyline.prototype._calcDimensions.call(this);
    const [sX, sY] = this._getStartPoints()
    const [cpX, cpY] = this._generateFakeControlPoint();
    const [eX, eY] = this._getEndPoints();
    const [a, b] = fabric.util.getBoundsOfCurve(this.points[0].x, this.points[0].y, cpX, cpY, cpX, cpY, this.points[2].x, this.points[2].y)

    return {
      left: a.x,
      top: a.y,
      width: b.x - a.x,
      height: b.y - a.y
    }
  }
```

I think that v6 has a solution to offer, using stroke projection somehow (with round stroke )
https://github.com/fabricjs/fabric.js/blob/70f955c3795f788ab1bfd6c5da4eec26788e3f7d/src/util/misc/projectStroke/index.ts#L13

**Reply by ShaMan123** [reply]
> I know that actually. Does fabric js provide any class for quadratic curves except `fabric.Path`?

No

**Reply by umtdemr** [reply]
I found a solution but I don't think this is efficient. I'm basically getting 40 points in the curve with the time value. After that, I'm just calculating the width and height. That's it. What do you think?

```js
_calcDimensions() {
    const {x: sX, y: sY} = this.points[0];
    const {x: cpX, y: cpY} = this.points[1];
    const {x: eX, y: eY} = this.points[2];

    const fakeCpX = 2*cpX - sX/2 - eX/2;
    const fakeCpY = 2*cpY - sY/2 - eY/2;
    
    // get 40 points in the curve to calculate dimensions
    const curvePoints = [];
    const step = 1/40;   
    for(let d=0; d<1.001; d+=step) {
        const curvePointX = this._getQBezierValue(d, sX, fakeCpX, eX);
        const curvePointY = this._getQBezierValue(d, sY, fakeCpY, eY);
        curvePoints.push({x: curvePointX, y: curvePointY});
    }

    let minXCurve = fabric.util.array.min(curvePoints, 'x') || 0;
    let minYCurve = fabric.util.array.min(curvePoints, 'y') || 0;
    let maxXCurve = fabric.util.array.max(curvePoints, 'x') || 0;
    let maxYCurve = fabric.util.array.max(curvePoints, 'y') || 0;
    const widthCurve = (maxXCurve - minXCurve);
    const heightCurve = (maxYCurve - minYCurve);

    return {
        left: minXCurve,
        top: minYCurve,
        width: widthCurve,
        height: heightCurve,
    }
},
```

**Reply by umtdemr** [reply]
And I found another issue when I activate object caching. If I draw a really long curve, when zoom-in, the line and the text are getting blurry even if I set strokeUniform to true.

**Reply by ShaMan123** [reply]
look at `noScaleCache`
